datafusion/expr/src/expr.rs

```diff
@@ -393,6 +398,258 @@ impl ScalarFunction {
     }
 }
 
+/// WindowFunction
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+pub enum WindowFunctionDefinition {
+    /// A built in aggregate function that leverages an aggregate function
+    AggregateFunction(aggregate_function::AggregateFunction),
+    /// A a built-in window function
+    BuiltInWindowFunction(BuiltInWindowFunction),
+    /// A user defined aggregate function
+    AggregateUDF(Arc<udaf::AggregateUDF>),
+    /// A user defined aggregate function
+    WindowUDF(Arc<WindowUDF>),
+}
+
+/// Find DataFusion's built-in window function by name.
+pub fn find_df_window_func(name: &str) -> Option<WindowFunctionDefinition> {
+    let name = name.to_lowercase();
+    // Code paths for window functions leveraging ordinary aggregators and
+    // built-in window functions are quite different, and the same function
+    // may have different implementations for these cases. If the sought
+    // function is not found among built-in window functions, we search for
+    // it among aggregate functions.
+    if let Ok(built_in_function) = BuiltInWindowFunction::from_str(name.as_str()) {
+        Some(WindowFunctionDefinition::BuiltInWindowFunction(
+            built_in_function,
+        ))
+    } else if let Ok(aggregate) =
+        aggregate_function::AggregateFunction::from_str(name.as_str())
+    {
+        Some(WindowFunctionDefinition::AggregateFunction(aggregate))
+    } else {
+        None
+    }
+}
+
+impl fmt::Display for BuiltInWindowFunction {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", self.name())
+    }
+}
+
+impl fmt::Display for WindowFunctionDefinition {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            WindowFunctionDefinition::AggregateFunction(fun) => fun.fmt(f),
+            WindowFunctionDefinition::BuiltInWindowFunction(fun) => fun.fmt(f),
+            WindowFunctionDefinition::AggregateUDF(fun) => std::fmt::Debug::fmt(fun, f),
+            WindowFunctionDefinition::WindowUDF(fun) => fun.fmt(f),
+        }
+    }
+}
+
+/// A [window function] built in to DataFusion
+///
+/// [window function]: https://en.wikipedia.org/wiki/Window_function_(SQL)
+#[derive(Debug, Clone, PartialEq, Eq, Hash, EnumIter)]
+pub enum BuiltInWindowFunction {
```

Comment:
Maybe you could put `BuiltInWindowFunction` in `datafusion/expr/src/built_in_window_function.rs` to mirror   `BuiltInFunction`?  
https://github.com/apache/arrow-datafusion/blob/main/datafusion/expr/src/built_in_function.rs

Reply:
@alamb @haohuaijin have I understood correctly?
- `BuiltinWindowFunction` goes into `datafusion/expr/src/built_in_window_function.rs`
- `WindowFunction` becomes `WindowFunctionDefinition` and goes into `datafusion/expr/src/expr.rs` with its trait implementations
- public functions in `window_function` stays there

Reply:
> * `BuiltinWindowFunction` goes into `datafusion/expr/src/built_in_window_function.rs`
> * `WindowFunction` becomes `WindowFunctionDefinition` and goes into `datafusion/expr/src/expr.rs` with its trait implementations

yes, you are correct.
> * public functions in `window_function` stays there

maybe those public functions also can move to `datafusion/expr/src/built_in_window_function.rs`. cc @alamb